File: 2022-06-craft-conf-languages/examples/parse-digit.c
#include <stdio.h>

int parse_digit(char ch) {
    if (ch >= '0' && ch <= '9') {
        return ch - '0';
    }
    return -1;
}

int parse_file(const char* contents) {
    int digit = parse_digit(contents[0]);
    if (digit == -1) {
        return -1;
    }
    return 0;
}

void use_file(const char* contents) {
    int rc = parse_file(contents);
    if (rc == -1) {
        printf("not a digit!\n");
    }
}

int main() {
    printf("%d\n", parse_digit('2'));
}

File: 2022-06-craft-conf-languages/examples/parse_digit.hs
import Data.Char

data InvalidDigit = InvalidDigit

parseDigit ch =
  if ch >= '0' && ch <= '9' then
    Right (ord ch - ord '0')
  else
    Left InvalidDigit

parseFile contents = do
  parseDigit (head contents)

useFile contents =
  case parseFile contents of
    Right _ -> pure ()
    Left _  -> print "not a digit"

main =
  case parseDigit '2' of
    Right digit -> print digit
    _           -> pure ()
